test/gtest/mha_helper.hpp

```diff
@@ -0,0 +1,468 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include "tensor_holder.hpp"
+#include "conv_tensor_gen.hpp"
+
+#include <hip_float8.hpp>
+#include <nlohmann/json.hpp>
+
+namespace test {
+namespace cpu {
+
+using float8 = miopen_f8::hip_f8<miopen_f8::hip_f8_type::fp8>;
+
+struct CPUMHATestCase
+{
+    // represents total number of seq len present in batch
+    size_t batch_size;
+    size_t sequence_length;
+    size_t num_heads;
+    size_t problem_dimension;
+    float drop_out_rate;
+
+    friend std::ostream& operator<<(std::ostream& os, const CPUMHATestCase& tc)
+    {
+        return os << "(batch_size: " << tc.batch_size << " num_heads:" << tc.num_heads
+                  << " sequence_length:" << tc.sequence_length
+                  << " problem_dimension:" << tc.problem_dimension
+                  << " drop_out_rate:" << tc.drop_out_rate << " )";
+    }
+};
+
+float GetF8Scaling(float max_val)
+{
+    constexpr float fp8_E4M3_max = 240.0f;
+
+    return fp8_E4M3_max / max_val;
+}
+
+template <typename T>
+T FindMax4D(const tensor<T>& max_of_tensor)
```

Comment:
A lot of these helper functions on tensors can benefit from variadic lambda parameters like this: 
```
max_of_tensor.for_each([&] (auto... indices) {
   maxVal = std::max(maxVal, max_of_tensor(indices...));
});
```